packages/server/modules/comments/services/notifications.ts

```diff
@@ -0,0 +1,101 @@
+import { CommentRecord } from '@/modules/comments/helpers/types'
+import { CommentsEvents, onCommentEvent } from '@/modules/comments/events/emitter'
+import { ensureCommentSchema } from '@/modules/comments/services/commentTextService'
+import type { JSONContent } from '@tiptap/core'
+import { iterateContentNodes } from '@/modules/core/services/richTextEditorService'
+import { publishNotification } from '@/modules/notifications/services/publication'
+import { difference, flatten } from 'lodash'
+import { NotificationType } from '@/modules/notifications/helpers/types'
+import { addStreamCommentMentionActivity } from '@/modules/activitystream/services/streamActivityService'
+
+function findMentionedUserIds(doc: JSONContent) {
+  const mentionedUserIds = new Set<string>()
+
+  for (const node of iterateContentNodes(doc)) {
+    if (node.type === 'mention') {
+      const uid = node.attrs?.id
+      if (uid) {
+        mentionedUserIds.add(uid)
+      }
+    }
+  }
+
+  return [...mentionedUserIds]
+}
+
+function collectMentionedUserIds(comment: CommentRecord): string[] {
+  if (!comment.text) return []
+
+  const { doc } = ensureCommentSchema(comment.text)
+  if (!doc) return []
+
+  return findMentionedUserIds(doc)
+}
+
+async function sendNotificationsForUsers(userIds: string[], comment: CommentRecord) {
+  const { id, streamId, authorId, parentComment } = comment
+  const threadId = parentComment || id
+
+  await Promise.all(
+    flatten(
+      userIds.map((uid) => {
+        return [
+          // Actually send out notification
+          publishNotification(NotificationType.MentionedInComment, {
+            targetUserId: uid,
+            data: {
+              threadId,
+              streamId,
+              authorId,
+              commentId: id
+            }
+          }),
+          // Create activity item
+          addStreamCommentMentionActivity({
+            streamId,
+            mentionAuthorId: authorId,
+            mentionTargetId: uid,
+            commentId: id,
+            threadId
+          })
+        ]
+      })
+    )
+  )
+}
+
+async function processNewCommentMentions(newComment: CommentRecord) {
```

Comment:
This could be merged into the function as below, with an optional previousComment argument that defaults to an empty list of previous mentions.

Reply:
true, i'll merge em